File: src/core/java/net/remmintan/mods/minefortress/core/interfaces/tasks/ITask.java
package net.remmintan.mods.minefortress.core.interfaces.tasks;

import com.mojang.datafixers.util.Pair;
import net.minecraft.util.math.BlockPos;
import net.remmintan.mods.minefortress.core.TaskType;
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.IWorkerPawn;

public interface ITask extends IBaseTask {

    TaskType getTaskType();
    boolean hasAvailableParts();
    ITaskPart getNextPart(IWorkerPawn colonist);
    void returnPart(Pair<BlockPos, BlockPos> partStartAndEnd);
    void finishPart(ITaskPart part, IWorkerPawn colonist);

    default void addFinishListener(Runnable listener) {}

}

File: src/core/java/net/remmintan/mods/minefortress/core/interfaces/tasks/ITask.kt
package net.remmintan.mods.minefortress.core.interfaces.tasks

import com.mojang.datafixers.util.Pair
import net.minecraft.util.math.BlockPos
import net.remmintan.mods.minefortress.core.TaskType
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.IWorkerPawn

interface ITask : IBaseTask {
    val taskType: TaskType
    fun hasAvailableParts(): Boolean
    fun getNextPart(colonist: IWorkerPawn): ITaskPart?
    fun returnPart(partStartAndEnd: Pair<BlockPos, BlockPos>)
    fun finishPart(part: ITaskPart, colonist: IWorkerPawn)
}
